### DISCUSSION TITLE
Search Plan Selection for Hybrid Queries (Vectors and Filters)

### DISCUSSION THREAD

**Discussion by aabylay** [milvus-io/milvus#33240] [Q&A and General discussion]
**Observation:**
For search queries with high selectivity filters on the HNSW index, the recall is consistently 1.0, regardless of index parameters (see Pic. 1). I expected this behavior to resemble queries with lower selectivity (see Pic. 2).

**Experimental Setup:**
I am working with a small dataset of 10k vectors, each with a dimensionality of 384. I have experimented with three different HNSW indexes built with the following parameters: sparse (m=4), default (m=16), and dense (m=100). For each combination of selectivity and k-NN, I ran 250 queries.

**Question:**
It seems possible that the query plan decides to use brute-force for high selectivity filters, but I am not sure. I could not find information on how the query execution plan is selected from the Milvus or Knowhere repo. Another potential explanation could be the high effectiveness of the HNSW graph traversal algorithm due to the small dataset size. Do you have any suggestions on how to determine the actual cause?

![image](https://github.com/milvus-io/milvus/assets/26962362/86ba4b90-d8ef-4a85-aa7d-cbbb28961772)
![image](https://github.com/milvus-io/milvus/assets/26962362/344257b7-d15c-4ce7-9958-ba35c9d1198c)

**Comment by xiaofan-luan** [comment]
I think the dataset might be too small and too easy so you get 100% recall

**Reply by aabylay** [reply]
This still doesn't clarify, why the recall on sparse HNSW index for 10% selectivity is far from being perfect.
I noticed that perfect recall is somehow achieved also in 6.3% selectivity on sparse index, but not for 10%.
Can you pls guide me to the file where the query execution plan selection is happening?
![image](https://github.com/milvus-io/milvus/assets/26962362/a127c462-4020-4c72-ab96-027593eb9448)

**Reply by xiaofan-luan** [reply]
when filtering rate is low, knowhere has a policy to fallback ann search to knn.
